### PR TITLE
deploy: scream for attention on certain changes

### PR DESCRIPTION
**Pull request by heerener**

### CODE REVIEW

File: notifications/apptainer_singularity.txt
Comment (matz-e) [inline notifications/apptainer_singularity.txt:1]: Will end up in `bluebrain/notifications` I hope?

Comment (heerener) [reply notifications/apptainer_singularity.txt:1]: It did now :-)

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -201,3 +201,20 @@ clean_merged_prs:
...
     DEPLOYMENT_BASE: $DEPLOYMENT_BASE
     GITHUB_PULL_REQUEST_ID: "merged"
+
+
+# some changes require a human to be notified. We'll do this through JIRA tickets
+notifications:
+  needs: [atomic_build]
+  image: python:3.11-bullseye
+  rules:
+    - if: '$CI_COMMIT_BRANCH == $DEPLOYMENT_DEFAULT_BRANCH'
```
Comment (matz-e) [inline .gitlab-ci.yml:211]: Sorry, I think it's not going to be _quite_ as simple as this.

If a dependency is changed, singularity would be rebuilt, too. That would require a trigger, too.

I'm wondering if we have something in the artifacts we could test for... :thinking:

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -201,3 +201,20 @@ clean_merged_prs:
...
+      changes:
+        - var/spack/repos/builtin/packages/apptainer/*
+        - var/spack/repos/builtin/packages/singularity/*
+        - var/spack/repos/builtin/packages/singularityce/*
+  script:
+    - pip install -r bluebrain/notifications/requirements.txt
+    - git log -n 3 >> bluebrain/notifications/apptainer_singularity.txt
+    - echo -e '{noformat}\n\nThank you!' >> bluebrain/notifications/apptainer_singularity.txt
```
Comment (matz-e) [inline .gitlab-ci.yml]: I'm wondering if we could use some formatting inside the template and add something like
```
{git_log}

Thank you!
```
there, then read the file and do `contents.format(git_log=subprocess.check_output(["git", ...])`?

Then it's all nicely contained in the Python script + template.

Would also allow to replace the git history beginning of the URL with some environment variable pointing to the Github repo

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -214,7 +214,21 @@ notifications:
...
+    - apt-get update
+    - apt-get install -y jq wget
+    - echo "Finding ID of the atomic build"
+    - ATOMIC_BUILD_ID=$(wget --header "Content-Type:application/json" --header "PRIVATE-TOKEN:${API_READ_TOKEN}" -O- "${CI_API_V4_URL}/projects/${CI_PROJECT_ID}/pipelines/${CI_PIPELINE_ID}/bridges" | jq '.[] | select(.name == "atomic_build") | .downstream_pipeline.id')
+    - echo "Atomic build pipeline ID: ${ATOMIC_BUILD_ID}"
+    - set +e
+      # TODO: API_READ_TOKEN should be replaced by one not tied to my user account
+    - wget --header "Content-Type:application/json" --header "PRIVATE-TOKEN:${API_READ_TOKEN}" -O- "${CI_API_V4_URL}/projects/${CI_PROJECT_ID}/pipelines/${ATOMIC_BUILD_ID}/test_report"
+    - |
+      wget --header "Content-Type:application/json" --header "PRIVATE-TOKEN:${API_READ_TOKEN}" -O- "${CI_API_V4_URL}/projects/${CI_PROJECT_ID}/pipelines/${ATOMIC_BUILD_ID}/test_report" | jq -e '.test_suites[] | select(.name == "externals") .test_cases[] | select(.classname == "singularity" or .classname == "singularityce" or .classname == "apptainer")'
```
Comment (matz-e) [inline .gitlab-ci.yml]: Final test: modify an irrelevant package and see if this gets triggered? I don't see anything that qualifies if the package should be a successful / skipped / failed test.

Comment (heerener) [reply .gitlab-ci.yml]: I was under the impression that the test just had to be present, but you're right, skipped tests should be ignored. Added that now.

https://bbpgitlab.epfl.ch/hpc/personal/heeren/cicd/-/pipelines/141632 demonstrates the jq syntax, experiments were done here: https://bbpgitlab.epfl.ch/hpc/personal/heeren/cicd/-/blob/fetch-tests-through-api/.gitlab-ci.yml (feel free to re-run the pipeline on branch `fetch-tests-through-api`, it doesn't take long).

**Comment by heerener** [comment]
https://bbpgitlab.epfl.ch/hpc/spack/-/pipelines/141634 modified neuron instead of singularityce; no notifications run was triggered.

All temporary changes have been undone again, we should be good to merge.

**Review by matz-e — APPROVED** [review]

**Pull request merged**